astroid/nodes/_base_nodes.py

```diff
@@ -220,3 +247,524 @@ def _elsed_block_range(
+            ]
+        else:
+            methods = [
+                OperatorNode._bin_op(left, binary_opnode, op, right, context),
+                OperatorNode._bin_op(
+                    right, binary_opnode, op, left, reverse_context, reverse=True
+                ),
+            ]
+
+        if (
+            PY310_PLUS
+            and op == "|"
+            and (
+                isinstance(left, (bases.UnionType, nodes.ClassDef))
+                or isinstance(left, nodes.Const)
+                and left.value is None
+            )
+            and (
+                isinstance(right, (bases.UnionType, nodes.ClassDef))
+                or isinstance(right, nodes.Const)
+                and right.value is None
+            )
+        ):
+            methods.extend([partial(OperatorNode._bin_op_or_union_type, left, right)])
+        return methods
+
+    @staticmethod
+    def _infer_binary_operation(
+        left: InferenceResult,
+        right: InferenceResult,
+        binary_opnode: nodes.AugAssign | nodes.BinOp,
+        context: InferenceContext,
+        flow_factory: GetFlowFactory,
+    ) -> Generator[InferenceResult, None, None]:
+        """Infer a binary operation between a left operand and a right operand.
+
+        This is used by both normal binary operations and augmented binary
+        operations, the only difference is the flow factory used.
+        """
+        from astroid import helpers  # pylint: disable=import-outside-toplevel
+
+        context, reverse_context = OperatorNode._get_binop_contexts(
+            context, left, right
+        )
+        left_type = helpers.object_type(left)
+        right_type = helpers.object_type(right)
+        methods = flow_factory(
+            left, left_type, binary_opnode, right, right_type, context, reverse_context
+        )
+        for method in methods:
+            try:
+                results = list(method())
+            except AttributeError:
+                continue
+            except AttributeInferenceError:
+                continue
+            except InferenceError:
+                yield util.Uninferable
+                return
+            else:
+                if any(isinstance(result, util.UninferableBase) for result in results):
+                    yield util.Uninferable
+                    return
+
+                if all(map(OperatorNode._is_not_implemented, results)):
+                    continue
+                not_implemented = sum(
+                    1 for result in results if OperatorNode._is_not_implemented(result)
+                )
+                if not_implemented and not_implemented != len(results):
+                    # Can't infer yet what this is.
+                    yield util.Uninferable
+                    return
+
+                yield from results
+                return
+
+        yield util.Uninferable
+
+
+class AttributeNode(NodeNG):
+    expr: NodeNG
```

Comment:
This feels like this class should also define a `postinit` to couple this annotation to something actually setting it.

Reply:
It's more of a mixin. The actual concrete classes do have postinit methods. Should I document this or raise some sort of TypeError when attempting to instantiate?

Reply:
Perhaps make them `ABC`? However, I don't really like mixins in this way. We actually spend quite some time removing them from `pylint` because the way this mixins work don't allow us to guarantee that they function at runtime. We can't enforce that `expr` is being set.. Would it be feasible to make them have a `postinit` that needs to be called?

Reply:
Happy to do it, but just for my own understanding, why would a postinit help? When would it ever be called, and how does it help a static type checker? I'm sure there is a reason, I just haven't worked with `postinit` in astroid yet.

Reply:
In this case it should probably be an `init`. Sorry for the confusion!

Reply:
If the mixins need to be robust against unintended instantiation, then they're more trouble then they're worth just for the DRY benefit. I'll just extract helper functions that can be shared.

Reply:
I also greatly dislike mixin for the sake of being dry because having to initialise instance in a particular order and not being able to use typecheck (becauseof course you can't guarantee the instanciation particular order and a type checker will rightfully warn you) is order if magnitude worse than whatever design problem mixin are trying to solve.